Fixes #38840 - Always verify auth token and expire session except for smart proxies

jeremylenz opened the pull request:
~~RH Cloud change: https://github.com/theforeman/foreman_rh_cloud/pull/1119~~

Note that the changed lines take place _after_ this
```
if !Setting[:restrict_registered_smart_proxies] || auth_smart_proxy(allowed_smart_proxies)
      set_admin_user
      return true
    end
```

so they should only apply to requests coming from users (not smart proxies).

### To test

Set idle timeout setting to 1 minute
Set your hammer to use sessions:
```
$ cat ~/.hammer/cli.modules.d/foreman.yml
:foreman:
  :use_sessions: true
```

Then run
```
hammer auth login basic
```

enter username/password

Wait >1 minute
`hammer organization list` -- should not list organizations. Should instead ask you for username/password.

jeremylenz committed:
Fixes #38840 - Always verify auth token and expire session
  for non-smart-proxy API consumers

parthaa commented:
Wondering if we should also add the `set_taxonomy` and `check_user_enabled`

ShimShtein reviewed (dismissed):
LGTM, but I didn't test it

jeremylenz commented:
Fixed tests.

jeremylenz commented:
@parthaa Extracted some logic to a new method to make things more understandable, and added the missing methods (Claude also caught one additional method that was being skipped.) Also added tests.

jeremylenz commented:
rebased

ShimShtein commented on app/controllers/concerns/foreman/controller/smart_proxy_auth.rb:
```diff
@@ -32,12 +32,22 @@ def require_smart_proxy_or_login(features = nil)
       return false
     end
     authorize
+    verify_authenticity_token
```
We can short-circuit here, and stop the filter process, if one method returns `false`:
``` ruby
authorize && verify_authenticity_token && set_taxonomy ...
```

jeremylenz replied:
Except I don't think it's good if `session_expiry` is conditional..

ShimShtein commented on app/controllers/concerns/foreman/controller/smart_proxy_auth.rb:
```diff
@@ -32,12 +32,22 @@ def require_smart_proxy_or_login(features = nil)
+  def require_user_login
     require_login
+    check_user_enabled
     unless User.current
```
We can return early if one of the first filters fails.
```suggestion
unless require_login && check_user_enabled && User.current
```

jeremylenz commented:
@ShimShtein Updated (Also had to fix the return value of Foreman::Controller::Authentication#check_user_enabled to get the short circuits to work right.)

parthaa commented on app/controllers/concerns/foreman/controller/smart_proxy_auth.rb:
```diff
@@ -32,12 +32,18 @@ def require_smart_proxy_or_login(features = nil)
+  end
+
+  def require_user_login
+    unless require_login && check_user_enabled && User.current
```
```suggestion
    unless require_login && User.current.present? && check_user_enabled
```

jeremylenz commented:
@parthaa updated

ShimShtein commented on app/controllers/concerns/foreman/controller/smart_proxy_auth.rb:
```diff
@@ -32,12 +32,18 @@ def require_smart_proxy_or_login(features = nil)
       return false
     end
-    authorize
+    authorize && verify_authenticity_token && set_taxonomy
```
You would like to return the `false` result in case this line fails I suppose

jeremylenz replied:
That would mean `session_expiry` is not called which was the original bug, no?

ShimShtein replied:
It will reject the call altogether. If a filter returns false, it will halt the filter chain and the action will not be executed. This was the behavior when the filters were defined with the `before_action` directive. I think we should mimic the same behavior here as well.

jeremylenz replied:
I changed it to
```
unless authorize && verify_authenticity_token && set_taxonomy
      return false
    end
```
and it reintroduced the original bug. It skips `session_expiry` and shows you the information.

jeremylenz replied:
Turns out `verify_authenticity_token` (from Rails) always returns nil. Had to move it to the front and short-circuited the rest, see if you're happy with that

jeremylenz replied:
Same for `session_expiry` and `update_activity_time` - the short circuiting does not work well here.

jeremylenz replied:
Learning more.. The filter chain only checks `false`, not `nil` values. So a `&&` short-circuit is not equivalent. We have to explicitly check for `false` everywhere.

jeremylenz committed:
Refs #38840 - only use trusted_hosts for loopback

jeremylenz committed:
Refs #38840 - short-circuit methods

jeremylenz committed:
Refs #38840 - only check false to truly emulate filter chain halting

jeremylenz committed:
Refs #38840 - add test for non-SSL loopbacks

parthaa reviewed (commented):
LGTM code wise. Thanks!

chris1984 reviewed (approved):
Tested and works fine, tested on a hammer devel box that is not local to the foreman server.
```bash
[vagrant@ip-10-0-167-15 ~]$ hammer location list
[Foreman] Username: admin
[Foreman] Password for admin:
---|------------------|------------------|------------
ID | TITLE            | NAME             | DESCRIPTION
---|------------------|------------------|------------
2  | Default Location | Default Location |
---|------------------|------------------|------------
[vagrant@ip-10-0-167-15 ~]$ date
Wed Oct 29 04:27:51 PM EDT 2025
$ date
Wed Oct 29 04:40:52 PM EDT 2025
[vagrant@ip-10-0-167-15 ~]$ hammer location list
[Foreman] Username:
```

ColeHiggins2 reviewed (approved):
Looks good to me :)
